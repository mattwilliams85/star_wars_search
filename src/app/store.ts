import { configureStore } from '@reduxjs/toolkit';
import searchReducer from '../components/Search/searchSlice';

export default configureStore({
  reducer: {
    search: searchReducer,
  }
});