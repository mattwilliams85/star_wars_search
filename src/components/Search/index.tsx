import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { useNavigate } from "react-router-dom";
import classnames from 'classnames';

import { update, loadStatus } from './searchSlice';
import { fetchMovieList } from '../../api/movies';
import { ReactComponent as SearchIcon } from '../../assets/images/icon-search.svg';
import { ReactComponent as CloseIcon } from '../../assets/images/icon-back.svg';
import styles from './styles.module.css';

interface Props {
  isOpen?: boolean;
}

const Search: React.FC<Props> = () => {
  const [searchInput, setSearchInput] = useState('');
  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    getMovies(true);
  }, []);

  const getMovies = async function (preventRedirect?: boolean) {
    if (!preventRedirect) navigate('/');
    dispatch(loadStatus(true));
    const result = await fetchMovieList(searchInput);
    dispatch(update(result));
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      getMovies();
    }
  };

  return (
    <div className={styles.searchWrap}>
      <input
        role="textbox"
        aria-label="search"
        placeholder='Search for Star Wars films'
        onChange={e => setSearchInput(e.target.value)}
        value={searchInput}
        onKeyDown={handleKeyDown}
        className={classnames(styles.search, {
          [styles.withInput]: searchInput
        })}
      />
      <SearchIcon className={classnames(styles.icon, styles.searchIcon)} onClick={() => { getMovies(); }} />
      {searchInput &&
        <CloseIcon className={classnames(styles.icon, styles.closeIcon)} onClick={() => setSearchInput('')} />
      }
    </div>
  );
};

export default Search;
