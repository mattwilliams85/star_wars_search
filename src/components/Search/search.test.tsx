// @ts-nocheck
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import store from '../../app/store';
import Search from './index';
import { BrowserRouter } from "react-router-dom";

const fetchMovieListResponse = {
  count: 1,
  query: 'abc123',
  results: [
    {
      title: 'The Phantom Menace',
      episode_id: 1
    }
  ]
};

describe('<Search />', () => {
  let searchTextBox: HTMLElement;
  let data: { query: string; };

  const setup = () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <Search />
        </Provider>
      </BrowserRouter>
    );

    searchTextBox = screen.getByRole('textbox', {
      name: /search/i,
    });
  };

  beforeEach(() => {
    data = {
      query: 'phantom'
    };
    jest.spyOn(window, 'fetch').mockResolvedValue({
      json: async () => (fetchMovieListResponse)
    });
  });

  test('should accept input', async () => {
    setup();
    const user = userEvent.setup();
    await user.clear(searchTextBox);
    await user.type(searchTextBox, data.query);
    expect(searchTextBox).toHaveValue(data.query);
  });

  test('should update search state successfully', async () => {
    setup();
    const user = userEvent.setup();
    await user.clear(searchTextBox);
    await user.type(searchTextBox, data.query);
    await user.keyboard('enter');
    const state = store.getState().search;

    expect(state.count).toEqual(1);
    expect(state.results[0].title).toEqual(fetchMovieListResponse.results[0].title);
  });

});