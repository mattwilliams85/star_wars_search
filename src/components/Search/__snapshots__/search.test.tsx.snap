// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component 1`] = `
<div>
  <div
    class="searchWrap"
  >
    <input
      class="search"
      placeholder="Search for Star Wars films"
      value=""
    />
    <svg
      class="icon searchIcon"
    >
      icon-search.svg
    </svg>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <div
    class="searchWrap"
  >
    <input
      class="search"
      placeholder="Search for Star Wars films"
      value=""
    />
    <svg
      class="icon searchIcon"
    >
      icon-search.svg
    </svg>
  </div>
</div>
`;
